src/DotVVM.Framework/Controls/DotvvmControl.cs

```diff
@@ -286,10 +286,32 @@ protected virtual void RenderControl(IHtmlWriter writer, IDotvvmRequestContext c
             if (RenderBeforeControl(in r, writer, context))
                 return;
 
-            AddAttributesToRender(writer, context);
-            RenderBeginTag(writer, context);
-            RenderContents(writer, context);
-            RenderEndTag(writer, context);
+            var renderAdapter = GetRenderAdapter(context);
+
+            var addAttributesToRenderAdapter = renderAdapter?.AddAttributesToRender;
+            if (addAttributesToRenderAdapter != null)
+                addAttributesToRenderAdapter(this, writer, context);
+            else
+                AddAttributesToRender(writer, context);
+
+            var renderBeginTagAdapter = renderAdapter?.RenderBeginTag;
+            if (renderBeginTagAdapter != null)
+                renderBeginTagAdapter(this, writer, context);
+            else
+                RenderBeginTag(writer, context);
+
+
+            var renderContentsAdapter = renderAdapter?.RenderContents;
+            if (renderContentsAdapter != null)
+                renderContentsAdapter(this, writer, context);
+            else
+                RenderContents(writer, context);
+
+            var renderEndTagAdapter = renderAdapter?.RenderEndTag;
+            if (renderEndTagAdapter != null)
+                renderEndTagAdapter(this, writer, context);
+            else
+                RenderEndTag(writer, context);
```

Comment:
This `RenderControl` method is virtual and some controls just override it and control the rendering on themselves. For these controls, this render adapter mechanism will never work, but this is private implementation detail of those controls that should not be observable from outside. This adapter breaks the encapsulation:

* Will make the behavior inconsistent across controls
* Will make any change in this implementation a breaking change
* Since the adapters will depend on the distribution of logic between those methods, moving anything around will also break compatibility
...
* Actually, even adding new properties to controls will start breaking code, because the adapters will not count with that. And new parts of the code base might depend on it :/